Error Plugin: @percy/cli: can't resolve reference /snapshot#/properties/name from id /config/static

**Possible bug introduced in beta 59?**

Renovate bot is trying to update to `@percy/cli to v1.0.0-beta.59 ` and our action that runs E2E tests and take screenshots is now falling with the following error:

```
(node:1969) Error Plugin: @percy/cli: can't resolve reference /snapshot#/properties/name from id /config/static
module: @oclif/config@1.17.0
task: runHook init
plugin: @percy/cli
root: /home/runner/work/montevideo/montevideo/node_modules/@percy/cli
```

The same action was running properly on beta.58

Our script just runs: `percy exec -- cypress run` (and some cypress env variables are used too with the --env flag)

Got the same issue with nightly Cypress E2E tests, Tests are now not executing due the Percy CLI crashing with the error

`percy exec -t 350 -- node scripts/cypress-ci.js "staging" "***" "***" "***" "***" "***" "***" "0" "***"
--
255 |  
256 | (node:351) Error Plugin: @percy/cli: can't resolve reference /snapshot#/properties/name from id /config/static
257 | module: @oclif/config@1.17.0
258 | task: runHook init
259 | plugin: @percy/cli
260 | root: /codebuild/output/src235357999/src/git-codecommit.us-east-1.amazonaws.com/v1/repos/cloudranger-v2-testing/Cypress/node_modules/@percy/cli
261 | See more details with DEBUG=*
262 | Error: can't resolve reference /snapshot#/properties/name from id
263 | /config/static
264 | npm ERR! code ELIFECYCLE
265 | npm ERR! errno 1
266 | npm ERR! cloudranger-v2-cypress@1.0.0 cypress:runCI: `percy exec -t 350 -- node scripts/cypress-ci.js "staging" "***" "***" "***" "***" "***" "***" "0" "***"`
267 | npm ERR! Exit status 1`

--- a/packages/cli-command/test/command.test.js
+++ b/packages/cli-command/test/command.test.js
@@ -1,9 +1,10 @@
 import logger from '@percy/logger/test/helpers';
 import PercyConfig from '@percy/config';
+import { configSchema } from '@percy/core/dist/config';
 import PercyCommand, { flags } from '../src';
 
 // add config schema to test discovery flags
-PercyConfig.addSchema(require('@percy/core/dist/config').schema);
+PercyConfig.addSchema(configSchema);
 
 describe('PercyCommand', () => {
   let results;

--- a/packages/cli-exec/src/hooks/init.js
+++ b/packages/cli-exec/src/hooks/init.js
@@ -1,8 +1,8 @@
 import PercyConfig from '@percy/config';
-import { schema, migration } from '@percy/core/dist/config';
+import * as CoreConfig from '@percy/core/dist/config';
 
 // ensures the core schema and migration is loaded
 export default function() {
-  PercyConfig.addSchema(schema);
-  PercyConfig.addMigration(migration);
+  PercyConfig.addSchema(CoreConfig.schemas);
+  PercyConfig.addMigration(CoreConfig.migration);
 }

--- a/packages/cli-snapshot/src/commands/snapshot.js
+++ b/packages/cli-snapshot/src/commands/snapshot.js
@@ -7,7 +7,7 @@ import logger from '@percy/logger';
 import globby from 'globby';
 import picomatch from 'picomatch';
 import YAML from 'yaml';
-import { schema } from '../config';
+import { configSchema } from '../config';
 import pkg from '../../package.json';
 
 // Throw a better error message for invalid urls
@@ -43,13 +43,13 @@ export class Snapshot extends Command {
     // static only flags
     files: flags.string({
       description: 'one or more globs matching static file paths to snapshot',
-      default: schema.static.properties.files.default,
+      default: configSchema.static.properties.files.default,
       percyrc: 'static.files',
       multiple: true
     }),
     ignore: flags.string({
       description: 'one or more globs matching static file paths to ignore',
-      default: schema.static.properties.ignore.default,
+      default: configSchema.static.properties.ignore.default,
       percyrc: 'static.ignore',
       multiple: true
     })

--- a/packages/cli-snapshot/src/config.js
+++ b/packages/cli-snapshot/src/config.js
@@ -1,7 +1,7 @@
 import { snapshotSchema } from '@percy/core/dist/config';
 
 // Config schema for static directories
-export const schema = {
+export const configSchema = {
   static: {
     type: 'object',
     additionalProperties: false,
@@ -66,6 +66,11 @@ export const snapshotListSchema = {
     }
   }]
 };
+
+export const schemas = [
+  configSchema,
+  snapshotListSchema
+];
 
 export function migration(config, { map, del }) {
   /* eslint-disable curly */

--- a/packages/cli-snapshot/src/hooks/init.js
+++ b/packages/cli-snapshot/src/hooks/init.js
@@ -1,12 +1,7 @@
 import PercyConfig from '@percy/config';
-import {
-  schema,
-  snapshotListSchema,
-  migration
-} from '../config';
+import * as SnapshotConfig from '../config';
 
 export default function() {
-  PercyConfig.addSchema(schema);
-  PercyConfig.addSchema(snapshotListSchema);
-  PercyConfig.addMigration(migration);
+  PercyConfig.addSchema(SnapshotConfig.schemas);
+  PercyConfig.addMigration(SnapshotConfig.migration);
 }

--- a/packages/cli-snapshot/src/hooks/init.js
+++ b/packages/cli-snapshot/src/hooks/init.js
@@ -1,7 +1,9 @@
 import PercyConfig from '@percy/config';
+import * as CoreConfig from '@percy/core/dist/config';
 import * as SnapshotConfig from '../config';
 
 export default function() {
+  PercyConfig.addSchema(CoreConfig.schemas);
   PercyConfig.addSchema(SnapshotConfig.schemas);
   PercyConfig.addMigration(SnapshotConfig.migration);
 }

--- a/packages/cli-upload/src/hooks/init.js
+++ b/packages/cli-upload/src/hooks/init.js
@@ -1,7 +1,7 @@
 import PercyConfig from '@percy/config';
-import { schema, migration } from '../config';
+import * as UploadConfig from '../config';
 
 export default function() {
-  PercyConfig.addSchema(schema);
-  PercyConfig.addMigration(migration);
+  PercyConfig.addSchema(UploadConfig.schema);
+  PercyConfig.addMigration(UploadConfig.migration);
 }

--- a/packages/config/src/validate.js
+++ b/packages/config/src/validate.js
@@ -73,10 +73,16 @@ export function getSchema(name) {
 
 // Adds schemas to the config schema's properties. The config schema is removed, modified, and
 // replaced after the new schemas are added to clear any compiled caches. Existing schemas are
-// removed and replaced as well. If a schema id is provided as the second argument, the schema
-// will be set independently and not added to config schema's properties.
+// removed and replaced as well. If a schema has an existing $id, the schema will not be added
+// as config schema properties.
 export function addSchema(schemas) {
-  if (isArray(schemas) || schemas.$id) {
+  if (isArray(schemas)) {
+    return schemas.map(addSchema);
+  }
+
+  if (schemas.$id) {
+    let { $id } = schemas;
+    if (ajv.getSchema($id)) ajv.removeSchema($id);
     return ajv.addSchema(schemas);
   }
 
@@ -90,7 +96,7 @@ export function addSchema(schemas) {
     ajv.addSchema(schema, $id);
   }
 
-  ajv.addSchema(config, '/config');
+  return ajv.addSchema(config, '/config');
 }
 
 // Resets the schema by removing all schemas and inserting a new default schema.

--- a/packages/config/test/index.test.js
+++ b/packages/config/test/index.test.js
@@ -118,6 +118,39 @@ describe('PercyConfig', () => {
         message: 'must be a string, received null'
       }]);
     });
+
+    it('can add schemas to replace existing schemas', () => {
+      PercyConfig.addSchema({
+        $id: 'foo',
+        type: 'string'
+      });
+
+      PercyConfig.addSchema({
+        $id: 'foo',
+        type: 'number'
+      });
+
+      expect(PercyConfig.validate('foo', 'foo')).toEqual([{
+        path: '',
+        message: 'must be a number, received a string'
+      }]);
+    });
+
+    it('can add multiple schemas at a time', () => {
+      PercyConfig.addSchema([{
+        foo: { type: 'string' }
+      }, {
+        bar: { $ref: '/config/foo' }
+      }]);
+
+      expect(PercyConfig.validate({
+        foo: 'bar',
+        bar: 100
+      })).toEqual([{
+        path: 'bar',
+        message: 'must be a string, received a number'
+      }]);
+    });
   });
 
   describe('.getDefaults()', () => {

--- a/packages/core/src/config.js
+++ b/packages/core/src/config.js
@@ -3,7 +3,7 @@ import PercyConfig from '@percy/config';
 import { merge } from '@percy/config/dist/utils';
 
 // Common config options used in Percy commands
-export const schema = {
+export const configSchema = {
   snapshot: {
     type: 'object',
     additionalProperties: false,
@@ -208,6 +208,13 @@ export const snapshotDOMSchema = {
     ...snapshotSchema.properties
   }
 };
+
+// Convinient reference for schema registration
+export const schemas = [
+  configSchema,
+  snapshotSchema,
+  snapshotDOMSchema
+];
 
 // Migration function
 export function migration(config, { map, del, log }) {

--- a/packages/core/src/index.js
+++ b/packages/core/src/index.js
@@ -2,12 +2,8 @@
 const { default: PercyConfig } = require('@percy/config');
 const CoreConfig = require('./config');
 
-PercyConfig.addSchema(CoreConfig.schema);
+PercyConfig.addSchema(CoreConfig.schemas);
 PercyConfig.addMigration(CoreConfig.migration);
-
-// used for per-snapshot validation
-PercyConfig.addSchema(CoreConfig.snapshotSchema);
-PercyConfig.addSchema(CoreConfig.snapshotDOMSchema);
 
 // Export the Percy class with commonjs compatibility
 module.exports = require('./percy').default;

